src/test/operators/join_index_test.cpp

```diff
@@ -0,0 +1,548 @@
+#include "../base_test.hpp"
+#include "gtest/gtest.h"
+
+#include "all_type_variant.hpp"
+#include "operators/join_index.hpp"
+#include "operators/table_scan.hpp"
+#include "operators/table_wrapper.hpp"
+#include "storage/chunk_encoder.hpp"
+#include "storage/index/adaptive_radix_tree/adaptive_radix_tree_index.hpp"
+#include "storage/index/group_key/composite_group_key_index.hpp"
+#include "storage/index/group_key/group_key_index.hpp"
+#include "storage/table.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+template <typename DerivedIndex>
+class JoinIndexTest : public BaseTest {
+ protected:
+  void SetUp() override {
+    // load and create the indexed tables
+    _table_wrapper_a = load_table_with_index("src/test/tables/int_float.tbl", 2);
+    _table_wrapper_b = load_table_with_index("src/test/tables/int_float2.tbl", 2);
+    _table_wrapper_c = load_table_with_index("src/test/tables/int_string.tbl", 4);
+    _table_wrapper_d = load_table_with_index("src/test/tables/string_int.tbl", 3);
+    _table_wrapper_e = load_table_with_index("src/test/tables/int_int.tbl", 4);
+    _table_wrapper_f = load_table_with_index("src/test/tables/int_int2.tbl", 4);
+    _table_wrapper_g = load_table_with_index("src/test/tables/int_int3.tbl", 4);
+    _table_wrapper_h = load_table_with_index("src/test/tables/int_int4.tbl", 4);
+    _table_wrapper_i = load_table_with_index("src/test/tables/int5.tbl", 1);
+    _table_wrapper_j = load_table_with_index("src/test/tables/int3.tbl", 1);
+    _table_wrapper_k = load_table_with_index("src/test/tables/int4.tbl", 1);
+    _table_wrapper_l = load_table_with_index("src/test/tables/int.tbl", 1);
+    _table_wrapper_m =
+        load_table_with_index("src/test/tables/aggregateoperator/groupby_int_1gb_0agg/input_null.tbl", 20);
+    _table_wrapper_n =
+        load_table_with_index("src/test/tables/aggregateoperator/groupby_int_1gb_1agg/input_null.tbl", 20);
+
+    // execute all TableWrapper operators in advance
+    _table_wrapper_a->execute();
+    _table_wrapper_b->execute();
+    _table_wrapper_c->execute();
+    _table_wrapper_d->execute();
+    _table_wrapper_e->execute();
+    _table_wrapper_f->execute();
+    _table_wrapper_g->execute();
+    _table_wrapper_h->execute();
+    _table_wrapper_i->execute();
+    _table_wrapper_j->execute();
+    _table_wrapper_k->execute();
+    _table_wrapper_l->execute();
+    _table_wrapper_m->execute();
+    _table_wrapper_n->execute();
+  }
+
+  std::shared_ptr<TableWrapper> load_table_with_index(const std::string& filename, const size_t chunk_size) {
+    auto table = load_table(filename, chunk_size);
+
+    // TODO(anyone): replace with EncodingType::Dictionary as soon as all index types support new compression
+    ChunkEncoder::encode_all_chunks(table, ColumnEncodingSpec{EncodingType::DeprecatedDictionary});
+
+    for (ChunkID chunk_id{0}; chunk_id < table->chunk_count(); ++chunk_id) {
+      auto chunk = table->get_chunk(chunk_id);
+
+      std::vector<ColumnID> columns{1};
```

Comment:
If I remove the index generation (or the part in the index join that deals with indexes), the tests still work. Or in other words: This does not really test anything.

Reply:
It does test that the join logic of the index join is not faulty. We assume that the index join utilizes the index, but currently there is no test explicitly guaranteeing that, this is true. However, in the join benchmarks it should show that the index is used.

If we were to test that the index is actually used, our approach would be to subclass an existing index with functionality to track whether it was used.
Do you want us to add this? Is there a better way?

Reply:
I just discussed this with @Bouncner: To understand why an IndexJoin was (not) performant, we think it makes sense to track how many chunk joins actually used an index. That information can both be visualized as part of the operator in the console and can be used by the tests, which could assert that the indexes were actually used.